### ISSUE TITLE
Framerate bug?

### ISSUE THREAD

**Issue by adfcf** [opentk/opentk#1537]
Hello. 
There is something weird happening with the frame rate; I put it to be 60 Hz, but it ends up being ~30 Hz. 

![demo1](https://user-images.githubusercontent.com/69996873/202823646-d54bcdc3-c43f-4202-9b77-224f02fe5b46.png)

### Repro steps
The code I used to test it:
```C#
using OpenTK.Windowing.Common;
using OpenTK.Windowing.Desktop;

namespace OpenTKIssue {

    internal class Program : GameWindow {
        public Program(GameWindowSettings gameWindowSettings, NativeWindowSettings nativeWindowSettings) : base(gameWindowSettings, nativeWindowSettings) {
            UpdateFrame += MyUpdate;
            RenderFrame += MyRender;
        }

        double timer;

        int frames, updates;

        private void MyRender(FrameEventArgs args) {
            ++frames;
        }

        private void MyUpdate(FrameEventArgs args) {
            timer += args.Time;
            if (timer >= 1.0) {
                Console.WriteLine("fps {0} | ups {1}", frames, updates);
                frames = 0;
                updates = 0;
                timer = 0;
            }
            ++updates;
        }

        static void Main() {
            GameWindowSettings settings = new() {
                RenderFrequency = 60.0,
                UpdateFrequency = 60.0
            };
            using Program test = new(settings, NativeWindowSettings.Default);
            test.Run();
        }

    }
}
```

### Related information 

* Operating system: Windows 10
* OpenTK Version: 4.7.5
* What runtime: .NET 6.0
* P.S.: This doesn't happen prior to version 4.7.2.

Edit: I might be missing something, but I've spent so much time on it already

**Comment by myblindy** [comment]
This is exactly what I'm seeing in the 5.0-pre8 branch as well, where the change log says it merged the 4.x branch.

https://github.com/opentk/opentk/issues/1534

**Comment by adfcf** [comment]
Thanks for telling me that! Looks like I wasn't doing anything wrong. 
The above code works as expected in versions prior to 4.7.2, so maybe, it started in there.

Probably, since Run() method was considerably modified:
- 4.7.1 Main Loop:
```
            while (GLFW.WindowShouldClose(WindowPtr) == false)
            {
                ProcessEvents();
                DispatchUpdateFrame();

                if (!IsMultiThreaded)
                {
                    DispatchRenderFrame();
                }
            }

            OnUnload();
```
- 4.7.2 Main loop:
```
            while (GLFW.WindowShouldClose(WindowPtr) == false)
            {
                double timeToNextUpdateFrame = DispatchUpdateFrame();

                double sleepTime = timeToNextUpdateFrame;
                if (!IsMultiThreaded)
                {
                    double timeToNextRenderFrame = DispatchRenderFrame();

                    sleepTime = Math.Min(sleepTime, timeToNextRenderFrame);
                }

                if (sleepTime > 0)
                {
                    Thread.Sleep((int)Math.Floor(sleepTime * 1000));
                }
            }
```
You hypothesized about Thread.Sleep being the cause, right? 
Miscalculation of sleepTime?

**Comment by myblindy** [comment]
Yeah, it looks similar to the change I did a few years ago when I used OpenTK in production and people at work complained about it using 100% of a core busy waiting the time between threads, but I implemented that sleep differently: I slept in small increments until I reached close to the time the next frame is expected, then I busy waited the last bit of time. 

This version is relinquishing the thread far too much and cutting heavily into the next frame timing.

**Comment by NogginBops** [comment]
I've just opened a PR that aims to improve on this issue. Schedulers on modern operating systems are a pain to deal with.

I'm not sure how testable the PR is, but if any of you want to give it a shot that would be great. Just note that to build OpenTK when checked out you need to run `build.cmd` for the initial build to work, then you should be able to build from within Visual Studio

**Comment by adfcf** [comment]
It works gracefully for me; stable 60 Hz, as intended, and low CPU usage. Thank you very much, @NogginBops!

**Comment by subchannel13** [comment]
Just curious, has anyone looked into [SpinWait](https://learn.microsoft.com/en-us/dotnet/api/system.threading.spinwait?view=net-7.0). It's supposed to look like a busy waiting loop from the outside while doing all sorts of "keeping CPU cold" from the inside.

**Comment by adfcf** [comment]
Hey, @subchannel13. I've never heard of it, so I can't tell how effectively that would be. Mr. Häger right above came with a good solution anyway. Thanks for showing up.

**Comment by NogginBops** [comment]
> Just curious, has anyone looked into [SpinWait](https://learn.microsoft.com/en-us/dotnet/api/system.threading.spinwait?view=net-7.0). It's supposed to look like a busy waiting loop from the outside while doing all sorts of "keeping CPU cold" from the inside.

I was thinking of using that struct but if I remember correctly when I tried it it was more inconsistent than the current solution. Maybe it can be used for the spin wait portion of the new code, I think I tried that but I don't remember the results. So it might be worth to try it again and document the results.

**Comment by NogginBops** [comment]
I've just merged #1538 which means that this issue should be fixed in the next release.
